### PR TITLE
Add setting to limit distance snapping to current time

### PR DESCRIPTION
**Pull request by peppy**
As discussed in https://github.com/ppy/osu/discussions/23815#discussioncomment-6124116.

Could probably do with test behaviour, but dunno. I want to rewrite the whole distance snapping classes as they are very hard to comprehend.

### CODE REVIEW

File: osu.Game/Screens/Edit/Compose/Components/CircularDistanceSnapGrid.cs [hunk 1]
```diff
@@ -98,9 +101,13 @@ public override (Vector2 position, double time) GetSnappedPosition(Vector2 posit
             if (travelLength < DistanceBetweenTicks)
                 travelLength = DistanceBetweenTicks;
 
+            if (LimitedDistanceSnap.Value)
+                travelLength = SnapProvider.DurationToDistance(ReferenceObject, editorClock.CurrentTime - ReferenceObject.GetEndTime());
```
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/CircularDistanceSnapGrid.cs:105]: This breaks with non-1.0x distance spacing.

https://github.com/ppy/osu/assets/20418176/fbd0beb5-a9bc-4082-aa2f-3f74f4887eb0

The snap doesn't snap to the brightest gridline.

Something like this appears to help:

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/CircularDistanceSnapGrid.cs b/osu.Game/Screens/Edit/Compose/Components/CircularDistanceSnapGrid.cs
index 2eec833832..63886e38eb 100644
--- a/osu.Game/Screens/Edit/Compose/Components/CircularDistanceSnapGrid.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/CircularDistanceSnapGrid.cs
@@ -101,12 +101,11 @@ public override (Vector2 position, double time) GetSnappedPosition(Vector2 posit
             if (travelLength < DistanceBetweenTicks)
                 travelLength = DistanceBetweenTicks;
 
-            if (LimitedDistanceSnap.Value)
-                travelLength = SnapProvider.DurationToDistance(ReferenceObject, editorClock.CurrentTime - ReferenceObject.GetEndTime());
-
-            // When interacting with the resolved snap provider, the distance spacing multiplier should first be removed
-            // to allow for snapping at a non-multiplied ratio.
-            float snappedDistance = SnapProvider.FindSnappedDistance(ReferenceObject, travelLength / distanceSpacingMultiplier);
+            float snappedDistance = LimitedDistanceSnap.Value
+                ? SnapProvider.DurationToDistance(ReferenceObject, editorClock.CurrentTime - ReferenceObject.GetEndTime())
+                // When interacting with the resolved snap provider, the distance spacing multiplier should first be removed
+                // to allow for snapping at a non-multiplied ratio.
+                : SnapProvider.FindSnappedDistance(ReferenceObject, travelLength / distanceSpacingMultiplier);
 
             double snappedTime = StartTime + SnapProvider.DistanceToDuration(ReferenceObject, snappedDistance);
 
```

but I'm not 100% confident so probably better you take a look first.

**Review by bdach — DISMISSED** [review]

File: osu.Game/Screens/Edit/Compose/Components/CircularDistanceSnapGrid.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Screens/Edit/Compose/Components/CircularDistanceSnapGrid.cs:105]: The fix looks solid 👍

File: osu.Game/Screens/Edit/Editor.cs [hunk 2]
```diff
@@ -337,6 +339,10 @@ private void load(OsuConfigManager config)
                                             new ToggleMenuItem(EditorStrings.AutoSeekOnPlacement)
                                             {
                                                 State = { BindTarget = editorAutoSeekOnPlacement },
+                                            },
+                                            new ToggleMenuItem(EditorStrings.LimitedDistanceSnap)
+                                            {
+                                                State = { BindTarget = editorLimitedDistanceSnap },
```
Comment (bdach) [inline osu.Game/Screens/Edit/Editor.cs:345]: One thing I'm not entirely sure about is this being in the core editor menu. Of all four official rulesets currently, two don't have (and never will have) a distance snap grid, the third (catch) is a _different_ sort of distance snap grid and this setting as implemented for osu! doesn't really make much sense there, and the fourth is osu! the ruleset. So this is a bit wonky.

If you're fine with this for now then probably so am I, but just voicing remaining concerns before I go clicking any buttons.

Comment (peppy) [reply osu.Game/Screens/Edit/Editor.cs:345]: It's pretty much a temporary thing yeah. Need to allow rulesets to populate menus or something at some point.

**Review by bdach — APPROVED** [review]

**Pull request merged**